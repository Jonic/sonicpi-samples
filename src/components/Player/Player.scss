@import '../../styles/component-helper';

.c-player {
}
