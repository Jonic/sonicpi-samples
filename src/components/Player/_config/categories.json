[
  "ambi",
  "bass",
  "bd",
  "drum",
  "elec",
  "guit",
  "loop",
  "misc",
  "perc",
  "sn",
  "tabla"
]
