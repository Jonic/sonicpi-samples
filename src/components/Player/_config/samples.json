[
  "ambi_choir",
  "ambi_dark_woosh",
  "ambi_drone",
  "ambi_glass_hum",
  "ambi_glass_rub",
  "ambi_haunted_hum",
  "ambi_lunar_land",
  "ambi_piano",
  "ambi_soft_buzz",
  "ambi_swoosh",
  "bass_dnb_f",
  "bass_drop_c",
  "bass_hard_c",
  "bass_hit_c",
  "bass_thick_c",
  "bass_trance_c",
  "bass_voxy_c",
  "bass_voxy_hit_c",
  "bass_woodsy_c",
  "bd_808",
  "bd_ada",
  "bd_boom",
  "bd_fat",
  "bd_gas",
  "bd_haus",
  "bd_klub",
  "bd_pure",
  "bd_sone",
  "bd_tek",
  "bd_zome",
  "bd_zum",
  "drum_bass_hard",
  "drum_bass_soft",
  "drum_cowbell",
  "drum_cymbal_closed",
  "drum_cymbal_hard",
  "drum_cymbal_open",
  "drum_cymbal_pedal",
  "drum_cymbal_soft",
  "drum_heavy_kick",
  "drum_roll",
  "drum_snare_hard",
  "drum_snare_soft",
  "drum_splash_hard",
  "drum_splash_soft",
  "drum_tom_hi_hard",
  "drum_tom_hi_soft",
  "drum_tom_lo_hard",
  "drum_tom_lo_soft",
  "drum_tom_mid_hard",
  "drum_tom_mid_soft",
  "elec_beep",
  "elec_bell",
  "elec_blip",
  "elec_blip2",
  "elec_blup",
  "elec_bong",
  "elec_chime",
  "elec_cymbal",
  "elec_filt_snare",
  "elec_flip",
  "elec_fuzz_tom",
  "elec_hi_snare",
  "elec_hollow_kick",
  "elec_lo_snare",
  "elec_mid_snare",
  "elec_ping",
  "elec_plip",
  "elec_pop",
  "elec_snare",
  "elec_soft_kick",
  "elec_tick",
  "elec_triangle",
  "elec_twang",
  "elec_twip",
  "elec_wood",
  "guit_e_fifths",
  "guit_e_slide",
  "guit_em9",
  "guit_harmonics",
  "loop_amen",
  "loop_amen_full",
  "loop_breakbeat",
  "loop_compus",
  "loop_garzul",
  "loop_industrial",
  "loop_mika",
  "loop_safari",
  "loop_tabla",
  "misc_burp",
  "misc_cineboom",
  "misc_crow",
  "perc_bell",
  "perc_snap",
  "perc_snap2",
  "perc_swash",
  "perc_till",
  "sn_dolf",
  "sn_dub",
  "sn_zome",
  "tabla_dhec",
  "tabla_ghe1",
  "tabla_ghe2",
  "tabla_ghe3",
  "tabla_ghe4",
  "tabla_ghe5",
  "tabla_ghe6",
  "tabla_ghe7",
  "tabla_ghe8",
  "tabla_ke1",
  "tabla_ke2",
  "tabla_ke3"
]