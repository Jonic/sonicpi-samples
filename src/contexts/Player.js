import React, { useEffect, useState } from 'react'

import PropTypes from 'prop-types'
import categoriesConfig from 'config/categories'
import { getSamplesForCategory } from 'helpers/SamplesCategories'
import samplesConfig from 'config/samples'

const PlayerContext = React.createContext({})

const PlayerContextProvider = ({ children }) => {
  const [samples, setSamples] = useState(samplesConfig)
  const [categories, setCategories] = useState(categoriesConfig)

  const [audioContext, setAudioContext] = useState(null)

  const [currentCategory, setCurrentCategory] = useState(categories[0])
  const [currentSample, setCurrentSample] = useState(samples[0])
  const [filteredSamples, setFilteredSamples] = useState(
    getSamplesForCategory(currentCategory),
  )

  const [isLooping, setIsLooping] = useState(false)
  const [isPlaying, setIsPlaying] = useState(false)

  useEffect(() => {
    let filteredSamplesForCategory = getSamplesForCategory(currentCategory)
    setFilteredSamples(filteredSamplesForCategory)
  }, [currentCategory])

  useEffect(() => {
    setCurrentSample(filteredSamples[0])
  }, [filteredSamples])

  useEffect(() => {
    setIsPlaying(false)
  }, [currentSample])

  const contextValue = {
    audioContext,
    categories,
    currentCategory,
    currentSample,
    filteredSamples,
    isLooping,
    isPlaying,
    samples,
    setAudioContext,
    setCategories,
    setCurrentCategory,
    setCurrentSample,
    setIsLooping,
    setIsPlaying,
    setSamples,
  }

  return (
    <PlayerContext.Provider value={contextValue}>
      {children}
    </PlayerContext.Provider>
  )
}

PlayerContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
}

export default PlayerContext
export { PlayerContextProvider }
