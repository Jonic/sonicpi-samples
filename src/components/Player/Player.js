import React, { Component } from 'react'
import PropTypes from 'prop-types'

import categories from './_config/categories'
import samples from './_config/samples'

import Audio from './Audio/Audio'
import Controls from './Controls/Controls'
// import Debug from './Debug/Debug'
// import Visualizer from './Visualizer/Visualizer'

class Player extends Component {
  constructor(props) {
    super(props)

    this.state = {
      bufferLength:    0,
      categories:      categories,
      currentCategory: categories[0],
      currentSample:   samples[0],
      filteredSamples: samples,
      isLooping:       false,
      isPlaying:       false,
      samples:         samples,
      visData:         null,
    }
  }

  getSamplesForCategory = category => {
    if (category === 'all') {
      return this.state.samples
    }

    let filteredSamples = []

    this.state.samples.map(sample => {
      if (sample.startsWith(category)) {
        filteredSamples.push(sample)
      }

      return null
    })

    return filteredSamples
  }

  loopCheckboxClickHandler = () => {
    this.setState({
      isLooping: !this.state.isLooping,
    })
  }

  playStopClickHandler = () => {
    this.toggleIsPlayingState()
  }

  categorySelectorChangeHandler = category => {
    let filteredSamples = this.getSamplesForCategory(category)

    this.setState({
      currentCategory: category,
      currentSample:   filteredSamples[0],
      filteredSamples: filteredSamples,
      isPlaying:       false,
    })
  }

  sampleEndedHandler = () => {
    this.setIsPlayingStateTo(false)
  }

  sampleSelectorChangeHandler = sample => {
    this.setState({
      currentSample: sample,
      isPlaying:     false,
    })
  }

  setIsPlayingStateTo = newState => {
    this.setState({
      isPlaying: newState,
    })
  }

  toggleIsPlayingState = () => {
    this.setIsPlayingStateTo(!this.state.isPlaying)
  }

  updateBufferLength = bufferLength => {
    this.setState({
      bufferLength: bufferLength,
    })
  }

  updateVisData = visData => {
    this.setState({
      visData: visData,
    })
  }

  render() {
    return (
      <div className="c-Player">
        <h1>Sonic Pi Samples</h1>

        {/* <Debug
          currentCategory={this.state.currentCategory}
          currentSample={this.state.currentSample}
          isLoading={this.props.isLoading}
          isLooping={this.state.isLooping}
          isPlaying={this.state.isPlaying}
        /> */}

        <Controls
          categories={this.state.categories}
          categorySelectorChangeHandler={this.categorySelectorChangeHandler}
          currentCategory={this.state.currentCategory}
          currentSample={this.state.currentSample}
          filteredSamples={this.state.filteredSamples}
          isLoading={this.props.isLoading}
          isLooping={this.state.isLooping}
          isPlaying={this.state.isPlaying}
          loopCheckboxClickHandler={this.loopCheckboxClickHandler}
          playStopClickHandler={this.playStopClickHandler}
          sampleSelectorChangeHandler={this.sampleSelectorChangeHandler}
          updateIsLoading={this.props.updateIsLoading}
        />

        <Audio
          currentSample={this.state.currentSample}
          isLooping={this.state.isLooping}
          isPlaying={this.state.isPlaying}
          sampleEndedHandler={this.sampleEndedHandler}
          updateIsLoading={this.props.updateIsLoading}
          updateBufferLength={this.updateBufferLength}
          updateVisData={this.updateVisData}
        />

        {/* <Visualizer
          bufferLength={this.state.bufferLength}
          isPlaying={this.state.isPlaying}
          visData={this.state.visData}
        /> */}
      </div>
    )
  }
}

Player.propTypes = {
  isLoading:       PropTypes.bool.isRequired,
  updateIsLoading: PropTypes.func.isRequired,
}

export default Player
