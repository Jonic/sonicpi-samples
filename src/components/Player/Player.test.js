import { shallow } from 'enzyme'
import categories from '../../_config/categories'
import Player from './Player'
import React from 'react'
import samples from '../../_config/samples'

const isLoading = false
const toggleIsLoading = jest.fn()
const updateIsLoading = jest.fn()

let wrapper

describe('<Player />', () => {
  beforeEach(() => {
    wrapper = shallow(
      <Player
        isLoading={isLoading}
        toggleIsLoading={toggleIsLoading}
        updateIsLoading={updateIsLoading}
      />
    )
  })

  describe('Renders...', () => {
    it('without crashing', () => {
      expect(wrapper.find('.c-Player')).toBePresent()
    })

    it('matches its snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    })
  })

  describe('Methods', () => {
    describe('categorySelectorChangeHandler()', () => {
      describe('when user selects "all"', () => {
        it('sets state correctly', () => {
          let category = 'all'
          let expectedCurrentCategory = categories[0]
          let expectedCurrentSample = samples[0]
          let expectedFilteredSamples = samples

          wrapper.instance().categorySelectorChangeHandler(category)

          let newCurrentCategory = wrapper.state().currentCategory
          let newCurrentSample = wrapper.state().currentSample
          let newFilteredSamples = wrapper.state().filteredSamples

          expect(newCurrentCategory).toEqual(expectedCurrentCategory)
          expect(newCurrentSample).toEqual(expectedCurrentSample)
          expect(newFilteredSamples).toEqual(expectedFilteredSamples)
        })
      })

      describe('when user selects "sn"', () => {
        it('sets state correctly', () => {
          let category = 'sn'
          let filteredSamples = [
            'sn_dolf',
            'sn_dub',
            'sn_zome',
          ]
          let expectedCurrentCategory = category
          let expectedCurrentSample = filteredSamples[0]
          let expectedFilteredSamples = filteredSamples

          wrapper.instance().categorySelectorChangeHandler(category)

          let newCurrentCategory = wrapper.state().currentCategory
          let newCurrentSample = wrapper.state().currentSample
          let newFilteredSamples = wrapper.state().filteredSamples

          expect(newCurrentCategory).toEqual(expectedCurrentCategory)
          expect(newCurrentSample).toEqual(expectedCurrentSample)
          expect(newFilteredSamples).toEqual(expectedFilteredSamples)
        })
      })
    })

    describe('loopCheckboxClickHandler', () => {
      it('toggles `isLooping` state value', () => {
        let isLooping = wrapper.state().isLooping
        wrapper.instance().loopCheckboxClickHandler()
        expect(wrapper.state().isLooping).toEqual(!isLooping)
      })
    })

    describe('playStopClickHandler', () => {
      it('toggles `isPlaying` state value', () => {
        let isPlaying = wrapper.state().isPlaying
        wrapper.instance().playStopClickHandler()
        expect(wrapper.state().isPlaying).toEqual(!isPlaying)
      })
    })

    describe('sampleEndedHandler', () => {
      it('sets `isPlaying` state value to `false`', () => {
        wrapper.instance().sampleEndedHandler()
        expect(wrapper.state().isPlaying).toBeFalsy()
      })
    })

    describe('sampleSelectorChangeHandler', () => {
      it('sets state correctly', () => {
        let expectedCurrentSample = samples[0]
        let expectedIsPlaying = false

        wrapper.instance().sampleSelectorChangeHandler(expectedCurrentSample)

        let currentSample = wrapper.state().currentSample
        let isPlaying = wrapper.state().isPlaying

        expect(currentSample).toEqual(expectedCurrentSample)
        expect(isPlaying).toEqual(expectedIsPlaying)
      })
    })

    describe('updateBufferLength', () => {
      it('sets state correctly', () => {
        let expectedBufferLength = 64
        wrapper.instance().updateBufferLength(expectedBufferLength)
        expect(wrapper.state().bufferLength).toEqual(expectedBufferLength)
      })
    })

    describe('updateVisData', () => {
      it('sets state correctly', () => {
        let expectedVisData = {
          foo: 'bar',
        }
        wrapper.instance().updateVisData(expectedVisData)
        expect(wrapper.state().visData).toEqual(expectedVisData)
      })
    })
  })
})
