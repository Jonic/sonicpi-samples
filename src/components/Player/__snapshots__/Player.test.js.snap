// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Player /> Renders... matches its snapshot 1`] = `
<div
  className="c-Player"
>
  <Debug
    currentCategory="all"
    currentSample="ambi_choir"
    isLoading={false}
    isLooping={false}
    isPlaying={false}
  />
  <Controls
    categories={
      Array [
        "all",
        "ambi",
        "bass",
        "bd",
        "drum",
        "elec",
        "guit",
        "loop",
        "misc",
        "perc",
        "sn",
        "tabla",
        "vinyl",
      ]
    }
    categorySelectorChangeHandler={[Function]}
    currentCategory="all"
    currentSample="ambi_choir"
    filteredSamples={
      Array [
        "ambi_choir",
        "ambi_dark_woosh",
        "ambi_drone",
        "ambi_glass_hum",
        "ambi_glass_rub",
        "ambi_haunted_hum",
        "ambi_lunar_land",
        "ambi_piano",
        "ambi_soft_buzz",
        "ambi_swoosh",
        "bass_dnb_f",
        "bass_drop_c",
        "bass_hard_c",
        "bass_hit_c",
        "bass_thick_c",
        "bass_trance_c",
        "bass_voxy_c",
        "bass_voxy_hit_c",
        "bass_woodsy_c",
        "bd_808",
        "bd_ada",
        "bd_boom",
        "bd_fat",
        "bd_gas",
        "bd_haus",
        "bd_klub",
        "bd_pure",
        "bd_sone",
        "bd_tek",
        "bd_zome",
        "bd_zum",
        "drum_bass_hard",
        "drum_bass_soft",
        "drum_cowbell",
        "drum_cymbal_closed",
        "drum_cymbal_hard",
        "drum_cymbal_open",
        "drum_cymbal_pedal",
        "drum_cymbal_soft",
        "drum_heavy_kick",
        "drum_roll",
        "drum_snare_hard",
        "drum_snare_soft",
        "drum_splash_hard",
        "drum_splash_soft",
        "drum_tom_hi_hard",
        "drum_tom_hi_soft",
        "drum_tom_lo_hard",
        "drum_tom_lo_soft",
        "drum_tom_mid_hard",
        "drum_tom_mid_soft",
        "elec_beep",
        "elec_bell",
        "elec_blip",
        "elec_blip2",
        "elec_blup",
        "elec_bong",
        "elec_chime",
        "elec_cymbal",
        "elec_filt_snare",
        "elec_flip",
        "elec_fuzz_tom",
        "elec_hi_snare",
        "elec_hollow_kick",
        "elec_lo_snare",
        "elec_mid_snare",
        "elec_ping",
        "elec_plip",
        "elec_pop",
        "elec_snare",
        "elec_soft_kick",
        "elec_tick",
        "elec_triangle",
        "elec_twang",
        "elec_twip",
        "elec_wood",
        "guit_e_fifths",
        "guit_e_slide",
        "guit_em9",
        "guit_harmonics",
        "loop_amen",
        "loop_amen_full",
        "loop_breakbeat",
        "loop_compus",
        "loop_garzul",
        "loop_industrial",
        "loop_mika",
        "loop_safari",
        "loop_tabla",
        "misc_burp",
        "misc_cineboom",
        "misc_crow",
        "perc_bell",
        "perc_snap",
        "perc_snap2",
        "perc_swash",
        "perc_till",
        "sn_dolf",
        "sn_dub",
        "sn_zome",
        "tabla_dhec",
        "tabla_ghe1",
        "tabla_ghe2",
        "tabla_ghe3",
        "tabla_ghe4",
        "tabla_ghe5",
        "tabla_ghe6",
        "tabla_ghe7",
        "tabla_ghe8",
        "tabla_ke1",
        "tabla_ke2",
        "tabla_ke3",
        "tabla_na",
        "tabla_na_o",
        "tabla_na_s",
        "tabla_re",
        "tabla_tas1",
        "tabla_tas2",
        "tabla_tas3",
        "tabla_te1",
        "tabla_te2",
        "tabla_te_m",
        "tabla_te_ne",
        "tabla_tun1",
        "tabla_tun2",
        "tabla_tun3",
        "vinyl_backspin",
        "vinyl_hiss",
        "vinyl_rewind",
        "vinyl_scratch",
      ]
    }
    isLoading={false}
    isLooping={false}
    isPlaying={false}
    loopCheckboxClickHandler={[Function]}
    playStopClickHandler={[Function]}
    sampleSelectorChangeHandler={[Function]}
    updateIsLoading={[Function]}
  />
  <Audio
    currentSample="ambi_choir"
    isLooping={false}
    isPlaying={false}
    sampleEndedHandler={[Function]}
    updateBufferLength={[Function]}
    updateIsLoading={[Function]}
    updateVisData={[Function]}
  />
  <Visualizer
    bufferLength={0}
    isPlaying={false}
    visData={null}
  />
</div>
`;
